### ISSUE TITLE
`scan_csv` uses enormous amount of memory on simple `head` call.

### ISSUE THREAD

**Issue by laserson** [pola-rs/polars#18724]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

Download this large CSV file from the OAS database:

https://opig.stats.ox.ac.uk/webapps/ngsdb/unpaired/Soto_2019/csv/SRR8365432_1_Heavy_IGHA.csv.gz

The gzipped file is about 2.7 GB.

Then run the following.

```python
import polars as pl
df = pl.scan_csv(
    "SRR8365432_1_Heavy_IGHA.csv.gz",
    has_header=True,
    skip_rows=1,
).head().collect()
```

My Python process starts using up all my RAM and then swapping. Sometimes the process gets killed by the OS. Same thing if I use `low_memory=True` in `scan_csv` or if I use `streaming=True` in `collect`.

This is very surprising to me that such a query would require all of my memory. Is this because even in streaming mode, Polars will materialize a big batch of rows in memory no matter what? Is it possible to control that? My goal would be to do some light casting of columns followed by `sink_parquet` with very little memory usage.

### Issue description

See above.

### Expected behavior

Low-memory usage for streaming a simple select => head or select => sink_parquet.

### Installed versions

<details>

```
--------Version info---------
Polars:              1.7.0
Index type:          UInt32
Platform:            macOS-14.3.1-arm64-arm-64bit
Python:              3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:54:21) [Clang 16.0.6 ]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2024.6.1
gevent               <not installed>
great_tables         <not installed>
matplotlib           3.8.3
nest_asyncio         1.6.0
numpy                1.26.4
openpyxl             3.1.5
pandas               2.2.1
pyarrow              17.0.0
pydantic             1.10.14
pyiceberg            <not installed>
sqlalchemy           2.0.28
torch                2.1.2.post3
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

**Comment by laserson** [comment]
And same thing if I first execute

```
pl.Config.set_streaming_chunk_size(10)
```

**Comment by laserson** [comment]
And finally, maybe a related issue, is that a call to `pl.read_csv_batched` also uses an enormous amount of memory before I even request any batches of data.

**Comment by aut0clave** [comment]
Does the issue occur if you uncompress the file prior to `scan_csv()`? I wonder if it's trying to uncompress the entire file into memory before applying the rest of the queryplan.

**Comment by laserson** [comment]
Yes, I can confirm the same behavior if I first uncompress the file.

**Comment by ritchie46** [comment]
We don't support streaming decompression yet. 

> Yes, I can confirm the same behavior if I first uncompress the file.

I cannot reproduce this. Slice push-down works on non-streaming files and runs in a fraction of the time in my local run.

**Comment by laserson** [comment]
> We don't support streaming decompression yet.

Is there a workaround for this? Perhaps by giving `scan_csv` a file-like object that handles the decompression?

> I cannot reproduce this. Slice push-down works on non-streaming files and runs in a fraction of the time in my local run.

Could you share the command that you are running? For me, the query does in fact run on the decompressed data, but it still uses a surprisingly large amount of memory.

**Comment by laserson** [comment]
As an alternative, could you recommend a separate way to convert the large gzipped csv file into Parquet without needing a huge amount of memory?

**Comment by natemcintosh** [comment]
In the meantime, I'd suggest duckdb for something like this. Its streaming engine is pretty well flushed out at this point. (That said, I don't know how well the streaming part works on compressed files).

If you have it [installed](https://duckdb.org/docs/installation/), then you can probably run something like

```sql
COPY 'big_file.csv.gz' TO 'data.parquet' (CODEC 'zstd');
```

If that still takes up all your RAM, you could try a few other things to help reduce RAM usage:
1. Set the number of bytes per file using the `FILE_SIZE_BYTES` parameter. Eg. setting it to 2GB file size would be `COPY 'big_file.csv.gz' TO 'data.parquet' (CODEC 'zstd', FILE_SIZE_BYTES 2000000000);`
1. Tell duckdb to limit its [RAM usage](https://duckdb.org/docs/configuration/overview.html#examples): `SET memory_limit = '10GB';` and then run again

Hope this helps.

**Comment by laserson** [comment]
Thanks @natemcintosh! Actually, I originally went with DuckDB first and when it failed on these jobs I picked up Polars. I was pretty surprised that DuckDB was dying even though these queries in principle require very little memory.

**Comment by natemcintosh** [comment]
That is very interesting! What was the main issue for duckdb? Running out of RAM?

**Comment by laserson** [comment]
Yes, similarly no matter what variation of the query/config I was trying to run, it always seemed to try to load the entire dataset into memory before anything, and then it would choke.

**Comment by natemcintosh** [comment]
That's a shame. I would guess the issue is that it has to decompress everything in memory, and can't stream the decompression. But that's only a hunch

**Comment by ghuls** [comment]
I managed to write a rough prototype that can stream gzipped files.

gzip compressed CSV file:
  - uncompressed size: 4,4GB
  - compressed size: 1.2 GB

Reading first 100000 rows of gzipped file with plain pl.read_csv ==> 12 GB RAM needed
```bash
❯ timeit ./polars_read_csv_head.py 
100000

Time output:
------------

  * Command: ./polars_read_csv_head.py
  * Elapsed wall time: 0:11.54 = 11.54 seconds
  * Elapsed CPU time:
     - User: 25.20
     - Sys: 6.11
  * CPU usage: 271%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 1130
     - Involuntarily (time slice expired): 14739
  * Maximum resident set size (RSS: memory) (kiB): 12018444
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 8
     - # of outputs: 0
  * Exit status: 0
```


Reading first 100000 rows of gzipped file with gzip streaming ==> 78 MB RAM needed
```
❯ timeit ./polars_read_streaming_csv_head.py 
100000

Time output:
------------

  * Command: ./polars_read_streaming_csv_head.py
  * Elapsed wall time: 0:00.16 = 0.16 seconds
  * Elapsed CPU time:
     - User: 0.15
     - Sys: 0.04
  * CPU usage: 126%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 282
     - Involuntarily (time slice expired): 23
  * Maximum resident set size (RSS: memory) (kiB): 67364
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0
```

Reading gzip compressed CSV file plain `pl.read_csv` and using all data of the file ==> 16GB RAM needed
```
❯ timeit ./polars_read_csv.py 

Time output:
------------

  * Command: ./polars_read_csv.py
  * Elapsed wall time: 0:15.67 = 15.67 seconds
  * Elapsed CPU time:
     - User: 51.30
     - Sys: 9.22
  * CPU usage: 386%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 1663
     - Involuntarily (time slice expired): 35483
  * Maximum resident set size (RSS: memory) (kiB): 16158804
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 0
     - # of outputs: 0
  * Exit status: 0
```

Reading gzip compressed CSV file with streaming approach and using all data to do some processing ==> 11GB RAM needed
```
❯ timeit ./polars_read_streaming_csv.py 

Time output:
------------

  * Command: ./polars_read_streaming_csv.py
  * Elapsed wall time: 0:15.14 = 15.14 seconds
  * Elapsed CPU time:
     - User: 52.54
     - Sys: 6.71
  * CPU usage: 391%
  * Context switching:
     - Voluntarily (e.g.: waiting for I/O operation): 91774
     - Involuntarily (time slice expired): 171934
  * Maximum resident set size (RSS: memory) (kiB): 11371272
  * Number of times the process was swapped out of main memory: 0
  * Filesystem:
     - # of inputs: 24
     - # of outputs: 0
  * Exit status: 0
```
